In [48]:
import pandas as pd
from sklearn.model_selection import KFold, cross_val_score
from sklearn.linear_model import LogisticRegression


SEED = 67

In [49]:
with open('Default.csv', 'r') as file:
    data = pd.read_csv(file)

print(f"First 5 rows of the dataset: \n{data.head()}\n")
print(f"Shape of the dataset: \n{data.shape}\n")
print(f"Null values: \n{data.isnull().sum()}\n")


First 5 rows of the dataset: 
   Unnamed: 0 default student      balance       income
0           1      No      No   729.526495  44361.62507
1           2      No     Yes   817.180407  12106.13470
2           3      No      No  1073.549164  31767.13895
3           4      No      No   529.250605  35704.49394
4           5      No      No   785.655883  38463.49588

Shape of the dataset: 
(10000, 5)

Null values: 
Unnamed: 0    0
default       0
student       0
balance       0
income        0
dtype: int64



In [50]:
# 10-fold CV
ten_fold_cv = KFold(n_splits=10, shuffle=True, random_state=SEED)

In [51]:
# Income and balance
X1 = data[['income', 'balance']]
y1 = (data['default'] == 'Yes').astype(int)

# Logistic regression model
lr_model1 = LogisticRegression(random_state=SEED)
cv_scores1 = cross_val_score(lr_model1, X1, y1, cv=ten_fold_cv, scoring='accuracy') # default 0.5 threshold 

print(f"10-fold CV test error with income and balance: \n{1 - cv_scores1.mean():.4f}")


10-fold CV test error with income and balance: 
0.0264


In [52]:
# Income, balance, and student
X2 = data[['income', 'balance', 'student']]
X2.loc[:, 'student'] = X2['student'].map({'Yes': 1, 'No': 0})
y2 = (data['default'] == 'Yes').astype(int)

# Logistic regression model
lr_model2 = LogisticRegression(random_state=SEED)
cv_scores2 = cross_val_score(lr_model2, X2, y2, cv=ten_fold_cv, scoring='accuracy')

print(f"10-fold CV test error with income, balance, and student: \n{1 - cv_scores2.mean():.4f}")

10-fold CV test error with income, balance, and student: 
0.0266


Part C:
Including a dummy variable for student status does NOT lead to a reduction of test error rate, since it increased from 2.64% (without student) to 2.66% (with student).